src/coreclr/jit/valuenum.cpp

```diff
@@ -7273,6 +7269,54 @@ void Compiler::fgValueNumberBlockAssignment(GenTree* tree)
                     isNewUniq = true;
                 }
 
+                if (!isNewUniq && (rhsVarDsc != nullptr) && (lhsVarDsc != nullptr))
+                {
+                    if (rhsLclVarTree->TypeGet() == TYP_STRUCT)
+                    {
+                        if (rhsLclVarTree->TypeGet() == lclVarTree->TypeGet())
+                        {
+                            assert(rhsVarDsc->TypeGet() == TYP_STRUCT);
+                            assert(lhsVarDsc->TypeGet() == TYP_STRUCT);
+
+                            CORINFO_CLASS_HANDLE rhsStructHnd = NO_CLASS_HANDLE;
+                            CORINFO_CLASS_HANDLE lhsStructHnd = NO_CLASS_HANDLE;
+
+                            if (rhsFldSeq == nullptr)
+                            {
+                                rhsStructHnd = rhsVarDsc->GetStructHnd();
+                            }
+                            else
+                            {
+                                CORINFO_FIELD_HANDLE fldHnd = rhsFldSeq->GetTail()->GetFieldHandle();
+                                rhsStructHnd                = info.compCompHnd->getFieldClass(fldHnd);
```

Comment:
There's a bit of a divergence between `doesFieldBelongToClass` and `getFieldClass` -- otherwise you would not have needed this new interface method. Can that cause problems here?